File: src/securityIdentity.ts
import registry from "../config/security-identities.json";

export type SecurityMarket = "us" | "hk" | "jp" | "kr" | "a" | "other";
export type SecurityIdentity = {
  code: string;
  name: string;
  market: SecurityMarket;
  marketLabel: string;
  exchange: string;
  aliases: string[];
  identityStatus: "verified" | "disclosed" | "pending";
};

const identityKey = (value: string) => value.trim().toUpperCase().replace(/[^A-Z0-9]/g, "");
const identities = new Map(registry.securities.flatMap((item) => item.aliases.map((alias) => [identityKey(alias), item] as const)));

/** Only explicitly verified aliases are merged; unknown disclosure codes stay intact. */
export function canonicalizeSecurityCode(code: string): string {
  return identities.get(identityKey(code))?.code ?? code.trim();
}

export function getSecurityIdentity(code: string, name = ""): SecurityIdentity {
  const item = identities.get(identityKey(code));
  if (item) {
    return {
      code: item.code,
      name: item.name,
      market: item.market as SecurityMarket,
      marketLabel: item.marketLabel,
      exchange: item.exchange,
      aliases: [...new Set([...item.aliases, ...item.nameAliases])],
      identityStatus: "verified",
    };
  }
  const normalized = code.trim().toUpperCase();
  let market: SecurityMarket = "other";
  let exchange = "交易所待核对";
  let identityStatus: SecurityIdentity["identityStatus"] = "pending";
  if (/^[A-Z0-9./-]+(?:US|UW|UN)(?:EQUITY)?$/.test(normalized)) {
    market = "us";
    exchange = "美国（披露市场标记）";
  } else if (/^\d{1,5}(?:HK|HG|HS)(?:EQUITY)?$/.test(normalized) || /^\d{1,5}\.HK$/.test(normalized)) {
    market = "hk";
    exchange = "香港（披露市场标记）";
  } else if (/^\d{5}$/.test(normalized)) {
    market = "hk";
    exchange = "HKEX";
    identityStatus = "disclosed";
  } else if (/^\d{6}[.]?(?:KS|KQ|KP)$/.test(normalized) || (/^\d{6}$/.test(normalized) && /SK\s*(?:HYNIX|海力)|三星|SAMSUNG|韩国|KOREA|现代汽车|起亚|LG|NAVER|Kakao|浦项|POSCO|Celltrion|韩华/i.test(name))) {
    market = "kr";
    exchange = "韩国（披露代码）";
  } else if (/^\d{4}\.(?:T|JP)$/.test(normalized) || (/^\d{4}$/.test(normalized) && /东京|丰田|索尼|日立|三菱|任天堂|软银|本田|东京电子|三井|住友|瑞穗|武田|迅销|基恩士|信越|村田|电装|佳能|尼康|日本/.test(name))) {
    market = "jp";
    exchange = "日本（披露代码）";
    identityStatus = "disclosed";
  } else if (/^[A-Z]{1,5}(?:[./-][A-Z]{1,2})?$/.test(normalized)) {
    market = "other";
    exchange = "交易所待核对";
    identityStatus = "pending";
  } else if (/^\d{6}$|^A\d+$/.test(normalized)) {
    market = "a";
    exchange = "中国内地（披露代码）";
    identityStatus = "disclosed";
  }
  const labels = { us: "美股", hk: "港股", kr: "韩股", jp: "日股", a: "A股", other: "市场待核对" };
  return { code: code.trim(), name: name || code.trim(), market, marketLabel: labels[market], exchange, aliases: [], identityStatus };
}

export function getSecurityMarket(code: string, name = ""): SecurityMarket {
  return getSecurityIdentity(code, name).market;
}

File: scripts/security_identity.py
"""Shared, evidence-bounded security identifiers for static data generation."""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any


CONFIG = json.loads((Path(__file__).resolve().parents[1] / "config/security-identities.json").read_text(encoding="utf-8"))


def identity_key(value: str) -> str:
    return re.sub(r"[^A-Z0-9]", "", value.strip().upper())


IDENTITIES = {identity_key(alias): item for item in CONFIG["securities"] for alias in item["aliases"]}


def verified_code_aliases() -> dict[str, str]:
    return {key: item["code"] for key, item in IDENTITIES.items()}


def canonicalize_security_code(code: str) -> str:
    item = IDENTITIES.get(identity_key(code))
    return item["code"] if item else code.strip()


def security_identity(code: str, name: str = "") -> dict[str, Any]:
    item = IDENTITIES.get(identity_key(code))
    if item:
        return {
            key: item[key] for key in ("code", "name", "market", "marketLabel", "exchange", "identityStatus")
        } | {"aliases": list(dict.fromkeys([*item["aliases"], *item.get("nameAliases", [])]))}
    normalized = code.strip().upper()
    market, exchange, status = "other", "交易所待核对", "pending"
    if re.fullmatch(r"[A-Z0-9./-]+(?:US|UW|UN)(?:EQUITY)?", normalized):
        market, exchange = "us", "美国（披露市场标记）"
    elif re.fullmatch(r"\d{1,5}(?:HK|HG|HS)(?:EQUITY)?", normalized) or re.fullmatch(r"\d{1,5}\.HK", normalized):
        market, exchange = "hk", "香港（披露市场标记）"
    elif re.fullmatch(r"\d{5}", normalized):
        market, exchange, status = "hk", "HKEX", "disclosed"
    elif re.fullmatch(r"\d{6}[.]?(?:KS|KQ|KP)", normalized) or (re.fullmatch(r"\d{6}", normalized) and re.search(r"SK\s*(?:HYNIX|海力)|三星|SAMSUNG|韩国|KOREA|现代汽车|起亚|LG|NAVER|Kakao|浦项|POSCO|Celltrion|韩华", name, re.I)):
        market, exchange = "kr", "韩国（披露代码）"
    elif re.fullmatch(r"\d{4}\.(?:T|JP)", normalized) or (re.fullmatch(r"\d{4}", normalized) and re.search(r"东京|丰田|索尼|日立|三菱|任天堂|软银|本田|东京电子|三井|住友|瑞穗|武田|迅销|基恩士|信越|村田|电装|佳能|尼康|日本", name)):
        market, exchange, status = "jp", "日本（披露代码）", "disclosed"
    elif re.fullmatch(r"[A-Z]{1,5}(?:[./-][A-Z]{1,2})?", normalized):
        market, exchange, status = "other", "交易所待核对", "pending"
    elif re.fullmatch(r"\d{6}|A\d+", normalized):
        market, exchange, status = "a", "中国内地（披露代码）", "disclosed"
    labels = {"us": "美股", "hk": "港股", "kr": "韩股", "jp": "日股", "a": "A股", "other": "市场待核对"}
    return {"code": code.strip(), "name": name or code.strip(), "market": market, "marketLabel": labels[market], "exchange": exchange, "aliases": [], "identityStatus": status}
